local-completions error with dataset longer than max_length

While testing meta-llama/Meta-Llama-3-8B model perpexity via vLLM, there is error on assessing context length. Here are steps to reproduce.
- start vLLM with model meta-llama/Meta-Llama-3-8B
- run evaluation
`lm-eval --model local-completions --tasks wikitext --model_args model=meta-llama/Meta-Llama-3-8B,base_url=http://localhost:8000/v1/completions,num_concurrent=1,tokenized_requests=True,max_length=8190 --verbosity DEBUG --log_samples --output_path test/`
It is causing an error on some requests:
```
  File "/lm-evaluation-harness/lm_eval/models/api_models.py", line 510, in _loglikelihood_tokens
    self.parse_logprobs(
  File "/lm-evaluation-harness/lm_eval/models/openai_completions.py", line 68, in parse_logprobs
    assert ctxlen > 0, "Context length must be greater than 0"
AssertionError: Context length must be greater than 0
```
It happens when len(context_enc) + len(continuation_enc) - (self.max_length) is equal to 1.

--- a/lm_eval/models/api_models.py
+++ b/lm_eval/models/api_models.py
@@ -104,7 +104,9 @@ class TemplateAPI(TemplateLM):
         self._truncate = truncate
         self._max_gen_toks = int(max_gen_toks)
         self._seed = int(seed)
-        self.max_length = max_length
+        # max_length - 1 as we always have 1 token for generation
+        eval_logger.info(f"Using max length {max_length} - 1")
+        self.max_length = max_length - 1
         if int(num_concurrent) <= 1:
             eval_logger.info(
                 "Concurrent requests are disabled. To enable concurrent requests, set `num_concurrent` > 1."
@@ -619,7 +621,8 @@ class TemplateAPI(TemplateLM):
                     utils.get_rolling_token_windows(
                         token_list=self.tok_encode(string),
                         prefix_token=self.prefix_token_id,
-                        max_seq_len=self.max_length,
+                        # max_seq_len - (1 for context)
+                        max_seq_len=self.max_length - 1,
                         context_len=1,
                     ),
                 )

--- a/lm_eval/models/api_models.py
+++ b/lm_eval/models/api_models.py
@@ -104,9 +104,8 @@ class TemplateAPI(TemplateLM):
         self._truncate = truncate
         self._max_gen_toks = int(max_gen_toks)
         self._seed = int(seed)
-        # max_length - 1 as we always have 1 token for generation
-        eval_logger.info(f"Using max length {max_length} - 1")
-        self.max_length = max_length - 1
+        eval_logger.info(f"Using max length {max_length}")
+        self.max_length = max_length
         if int(num_concurrent) <= 1:
             eval_logger.info(
                 "Concurrent requests are disabled. To enable concurrent requests, set `num_concurrent` > 1."
@@ -419,9 +418,10 @@ class TemplateAPI(TemplateLM):
         cache_keys = []
         for chunk in chunks:
             for cache_key, context_enc, continuation_enc in chunk:
-                inp = (context_enc + continuation_enc)[-(self.max_length) :]
+                # max_length - 1 as we always have 1 token for generation
+                inp = (context_enc + continuation_enc)[-(self.max_length - 1) :]
                 ctxlen = len(context_enc) - max(
-                    0, len(context_enc) + len(continuation_enc) - (self.max_length)
+                    0, len(context_enc) + len(continuation_enc) - (self.max_length - 1)
                 )
 
                 inputs.append(inp)

--- a/lm_eval/models/vllm_causallms.py
+++ b/lm_eval/models/vllm_causallms.py
@@ -289,7 +289,7 @@ class VLLM(TemplateLM):
                     make_disjoint_window,
                     get_rolling_token_windows(
                         token_list=self.tok_encode(string),
-                        prefix_token=self.eot_token_id,
+                        prefix_token=self.prefix_token_id,
                         max_seq_len=self.max_length - 1,
                         context_len=1,
                     ),

--- a/lm_eval/models/vllm_causallms.py
+++ b/lm_eval/models/vllm_causallms.py
@@ -290,6 +290,7 @@ class VLLM(TemplateLM):
                     get_rolling_token_windows(
                         token_list=self.tok_encode(string),
                         prefix_token=self.prefix_token_id,
+                        # max_seq_len - (1 for context)
                         max_seq_len=self.max_length - 1,
                         context_len=1,
                     ),

